src/buildercore/cfngen.py

```diff
@@ -109,108 +120,169 @@ def build_context(pname, **more_context): # pylint: disable=too-many-locals
         #}
     }
 
-    context = copy.deepcopy(defaults)
+    context = deepcopy(defaults)
     context.update(more_context)
 
+    ensure('stackname' in context, "'stackname' not provided") # this still sucks
+
     # proceed with wrangling
 
-    # TODO: don't like this. if a stackname is required, make it a parameter.
-    # stackname used to be derived inside this func from pname + id + cluster number
-    ensure(context['stackname'], "a stackname wasn't provided.")
-    stackname = context['stackname']
+    # why shuffle? so we don't have a context builder depending
+    # on another context builder mutating the context in a specific way
+    # wrangler_list = shuffle([
+    wrangler_list = [
+        partial(build_context_rds, existing_context=existing_context),
+        build_context_aws,
+        build_context_ec2,
+        build_context_elb,
+        build_context_cloudfront,
+        build_context_sns_sqs,
+        build_context_s3,
+        build_context_cloudfront,
+        build_context_fastly,
+        build_context_gcp,
+        build_context_subdomains,
+        build_context_elasticache,
+        build_context_vault,
+    ]
+
+    # exceptions to the rule ...
+    wrangler_list = [project_wrangler] + wrangler_list
+
+    for wrangler in wrangler_list:
+        # so functions can't modify the context in-place or
+        # reference a bit of project_data and then change it
+        context = wrangler(deepcopy(project_data), deepcopy(context))
```

Comment:
that's good, true immutability. Ideally they wouldn't need `context` as an input, right? But it would be difficult to merge them all.

Reply:
`project_name` and `stackname` are the only fields that are read from the context in this module, so I'd try to simplify that to remove `deepcopy(context)` as an argument. Would be a follow-up when this has been integrated.

Reply:
I can see pros and cons, but even though I commented it out, I think shuffling the order of the wranglers is a good idea to prevent ordering dependencies. Removing the context as a parameter would obviate the shuffling.